### PR TITLE
Add initial Pyvista visualization

### PR DESCRIPTION
**Pull request by dyvasey**
This adds a function `plot2D` that takes a Pyvista mesh and plots it using `imshow` in Matplotlib. In addition to the basic function in the module `visualization`, this pull request also includes a test `test_visualization.py` and a example Jupyter Notebook `visualization.ipynb`. There are also updates to the documentation to include the API of this new module/function, as well as updates to the setup files and files needed for Github Actions and Binder.

### CODE REVIEW

**Comment by dyvasey** [comment]
Hmmm, ok, all the tests failed for this. Will need to troubleshoot.

**Comment by dyvasey** [comment]
This is strange, because the tests run successfully on my local systems.

**Comment by dyvasey** [comment]
Alright! Finally figured out the problem is related to headless rendering. Luckily pyvista has a github workflow for setting this up, so all I had to do was add that to the yml. I will make a couple changes and then clean up the commit history with rebase.

**Comment by dyvasey** [comment]
Ok, the commit history is cleaned up and the checks are passing, so this is now ready for review.

File: apt.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+libgl1-mesa-glx
```
Comment (naliboff) [inline apt.txt:1]: Add a comment here that these are the packages for binder that are needed to run the headless display?

File: gdmate/analysis/visualization.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
+"""
+Module for visualizing model results using Pyvista and other tools
+"""
+import numpy as np
+import matplotlib.pyplot as plt
+import pyvista as pv
+
+def plot2D(mesh,field,bounds=None,ax=None,colorbar=False,**kwargs):
```
Comment (naliboff) [inline gdmate/analysis/visualization.py:8]: call this plot_pyvista_2d?

File: gdmate/analysis/visualization.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    xmag = float(abs(bounds_array[1] - bounds_array[0]))
+    ymag = float(abs(bounds_array[3] - bounds_array[2]))
+    aspect_ratio = ymag/xmag
+    
+    # Set a standard plotter window size
+    plotter.window_size = (1024,int(1024*aspect_ratio))
+    
+    xmid = xmag/2 + bounds_array[0] # X midpoint
+    ymid = ymag/2 + bounds_array[2] # Y midpoint
+    zoom = xmag*aspect_ratio*1.875 # Zoom level - not sure why 1.875 works
```
Comment (naliboff) [inline gdmate/analysis/visualization.py:51]:
```suggestion
    zoom = xmag*aspect_ratio*1.875 # Zoom level - 1.875 was determined via trial and error
```

File: gdmate/analysis/visualization.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    # Calculate Camera Position from Bounds
+    bounds_array = np.array(bounds)
+    xmag = float(abs(bounds_array[1] - bounds_array[0]))
+    ymag = float(abs(bounds_array[3] - bounds_array[2]))
+    aspect_ratio = ymag/xmag
+    
+    # Set a standard plotter window size
+    plotter.window_size = (1024,int(1024*aspect_ratio))
+    
+    xmid = xmag/2 + bounds_array[0] # X midpoint
```
Comment (naliboff) [inline gdmate/analysis/visualization.py:49]:
```suggestion
       # Define the X/Y midpoints, and zoom level. The ideal zoom factor of 1.875 was
       # determined by trial and error.
        xmid = xmag/2 + bounds_array[0] # X midpoint
```

File: gdmate/analysis/visualization.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    zoom = xmag*aspect_ratio*1.875 # Zoom level - not sure why 1.875 works
+    
+    # Set camera for plotter window
+    position = (xmid,ymid,zoom)
+    focal_point = (xmid,ymid,0)
+    viewup = (0,1,0)
+    
+    camera = [position,focal_point,viewup]
+    
+    plotter.camera_position = camera
```
Comment (naliboff) [inline gdmate/analysis/visualization.py:60]: Add a comment here.

File: tests/test_visualization.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+# Create very small mesh of 9 cells with 16 points and assign each a value
+mesh = pv.UniformGrid(dims=(4, 4, 1)).cast_to_unstructured_grid()
+mesh['sample_field'] = np.arange(9)
+
+def test_plot2D():
+    """ Test plot2D function """
+    fig,ax = plt.subplots(1)
+    ax = visualization.plot2D(mesh,'sample_field',bounds=[0,2,0,2],ax=ax)
+
+    # Test that the axes contains an image
```
Comment (naliboff) [inline tests/test_visualization.py:19]: Possible to add more asserts here (sans image comparisons)? The more the better for catching errors down the line.

File: notebooks/visualization.ipynb [hunk 7]
```diff
@@ -0,0 +1,89 @@
+{
```
Comment (gdmate) [inline notebooks/visualization.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    mesh = pv.UniformGrid(dims=(4, 4, 1)).cast_to_unstructured_grid()
</pre><p>Add comments above each line?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gdmate/gdmate/pull/20/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

**Review by naliboff — CHANGES_REQUESTED** [review]
Looks good, only a few minor changes!

**Comment by dyvasey** [comment]
Ok, I made the changes requested by the previous review and resolved a small conflict between the previous merged PR and this one.

File: apt.txt (hunk 1 above)
Comment (dyvasey) [reply apt.txt:3]: Added to file.

File: gdmate/analysis/visualization.py (hunk 2 above)
Comment (dyvasey) [reply gdmate/analysis/visualization.py]: Changed here and elsewhere.

File: gdmate/analysis/visualization.py (hunk 4 above)
Comment (dyvasey) [reply gdmate/analysis/visualization.py]: Moved inline comments to block above the code

File: gdmate/analysis/visualization.py (hunk 5 above)
Comment (dyvasey) [reply gdmate/analysis/visualization.py:67]: Added and removed `plotter.camera_set`, which appears to have been unnecessary.

File: notebooks/visualization.ipynb (hunk 7 above)
Comment (dyvasey) [reply notebooks/visualization.ipynb:1]: Added additional comments

File: tests/test_visualization.py (hunk 6 above)
Comment (dyvasey) [reply tests/test_visualization.py]: Added additional asserts that the figure contains an axes and for the y bounds as well.

**Review by naliboff — APPROVED** [review]
@dyvasey - I have a few additional changes in mind to this PR, but for now let's merge for the purpose of upcoming demonstrations and handle changes in a follow-up PR.

Changes that should be done in follow-up PRs include:
1. Move the primary functions to a folder called "visualization"
2. Improve variable/function documentation

**Pull request merged**